### PR TITLE
channeldb+routing+discovery: reject zombie announcements

### PR DESCRIPTION
**Pull request by wpaulino**
In this PR, we introduce a new persistent index for zombie edges to quickly reject announcements for edges we've previously deemed as zombies. Care has been taken to ensure we don't reject fresh updates for edges we've considered zombies.

### CODE REVIEW

File: discovery/gossiper.go [hunk 1]
```diff
@@ -1787,9 +1803,37 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
...
+				nMsg.err <- fmt.Errorf("rejecting "+
+					"ChannelUpdate(%v) for zombie channel",
+					msg.ShortChannelID)
+				return nil
+			}
+
+			// Otherwise, we'll mark the edge as live within the
+			// router to ensure we can process any future updates
+			// for this edge until it is considered zombie again.
+			err := d.cfg.Router.MarkEdgeLive(msg.ShortChannelID)
```
Comment (cfromknecht) [inline discovery/gossiper.go]: we should defer marking an edge as live only after we are sure  the signature validates, otherwise someone could blast updates w/ fake signatures that bring about a zombie uprising

File: discovery/gossiper.go [hunk 2]
```diff
@@ -1787,9 +1803,37 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
 		}
 
 		// Before we perform any of the expensive checks below, we'll
-		// make sure that the router doesn't already have a fresher
-		// announcement for this edge.
+		// make sure that we don't process any messages for known zombie
+		// edges.
 		timestamp := time.Unix(int64(msg.Timestamp), 0)
+		if d.cfg.Router.IsZombieEdge(msg.ShortChannelID) {
```
Comment (cfromknecht) [inline discovery/gossiper.go]: similar comment here with unifying `IsZombieEdge` w/ `IsStaleEdgePolicy` to avoid adding another db tx

File: routing/router.go [hunk 3]
```diff
@@ -435,6 +423,29 @@ func (r *ChannelRouter) Start() error {
 		}
 	}
 
+	// If AssumeChannelValid is present, then we'll only prune channels once
+	// we detect them as zombies.
```
Comment (cfromknecht) [inline routing/router.go]: comment doesn't seem to match code, this looks like we prune them immediately

Comment (cfromknecht) [reply routing/router.go]: i see now, would update comment to say if assume channel valid is present, we won't start a chain view to track spentness or something

Comment (wpaulino) [reply routing/router.go]: Fixed.

File: discovery/gossiper.go (hunk 2 above)
Comment (wpaulino) [reply discovery/gossiper.go]: Fixed.

File: discovery/gossiper.go (hunk 1 above)
Comment (wpaulino) [reply discovery/gossiper.go]: Since the `GetChannelByID` call below will fail due to it being a zombie channel, we'll go ahead and stash it so that it gets reprocessed once we receive its corresponding channel announcement. Without the announcement, we're unable to verify signatures since we don't have the public keys of each edge. If and when we receive the announcement, we won't be able to re-process it because it'll still be marked as a zombie.

File: discovery/gossiper.go [hunk 4]
```diff
@@ -1571,8 +1579,16 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
 			return nil
 		}
 
-		// At this point, we'll now ask the router if this is a stale
-		// update. If so we can skip all the processing below.
+		// At this point, we'll now ask the router if this is a
+		// zombie/known edge. If so we can skip all the processing
+		// below. In the case of a zombie edge, we'll reject it
+		// completely.
+		if d.cfg.Router.IsZombieEdge(msg.ShortChannelID) {
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply discovery/gossiper.go]: Fixed.

File: channeldb/graph.go [hunk 5]
```diff
@@ -2778,6 +2784,68 @@ func (c *ChannelGraph) NewChannelEdgePolicy() *ChannelEdgePolicy {
...
+		return zombieIndex.Delete(k[:])
+	})
+}
+
+// IsZombieEdge returns whether the edge is considered zombie.
+func (c *ChannelGraph) IsZombieEdge(chanID uint64) bool {
+	isZombie := false
+
+	// It's safe to ignore the error being returned, if any, as we'll just
+	// return false anyway.
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply channeldb/graph.go]: Fixed.

File: routing/router.go [hunk 6]
```diff
@@ -734,10 +734,17 @@ func (r *ChannelRouter) pruneZombieChans() error {
...
 			chanToPrune)
 
-		err := r.cfg.Graph.DeleteChannelEdge(&chanToPrune)
+		err := r.cfg.Graph.MarkEdgeZombie(chanToPrune.ChannelID)
+		if err != nil {
+			return fmt.Errorf("unable to mark ChannelPoint(%v) "+
+				"as zombie: %v", chanToPrune.ChannelPoint, err)
+		}
+
+		err = r.cfg.Graph.DeleteChannelEdge(&chanToPrune.ChannelPoint)
```
Comment (Roasbeef) [inline routing/router.go]: Perhaps we should combine both operations into a single call? If we do, then this means that someone can't give us a super old channel and make us re-scan the graph with GetUTXO or the like.

File: routing/router.go [hunk 7]
```diff
@@ -433,6 +421,30 @@ func (r *ChannelRouter) Start() error {
 		}
 	}
 
+	// If AssumeChannelValid is present, then we won't rely on pruning
+	// channels from the graph based on their spentness, but whether they
+	// are considered zombies or not.
+	if r.cfg.AssumeChannelValid {
+		if err := r.pruneZombieChans(); err != nil {
```
Comment (Roasbeef) [inline routing/router.go]: This skips calling `PruneGraphNodes`.

File: routing/router.go (hunk 6 above)
Comment (cfromknecht) [reply routing/router.go]: yeah the more i think about this, i think its a good idea to have the transition from live -> zombie be atomic

File: routing/router.go (hunk 7 above)
Comment (wpaulino) [reply routing/router.go]: Will be including these changes in another PR.

File: routing/router.go (hunk 6 above)
Comment (wpaulino) [reply routing/router.go]: Good point! Fixed.

File: discovery/gossiper_test.go [hunk 8]
```diff
@@ -2196,6 +2198,233 @@ func TestForwardPrivateNodeAnnouncement(t *testing.T) {
...
+	)
+
+	select {
+	case <-ctx.broadcastedMessage:
+		t.Fatal("expected to not broadcast live channel update " +
+			"without announcement")
+	case <-time.After(2 * trickleDelay):
+	}
+
+	time.Sleep(time.Second)
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply discovery/gossiper_test.go]: Since we have to re-process it, if we return to the caller immediately then they won't be informed of any errors that occurred while re-processing it. Turns out the sleep wasn't needed though.

File: routing/router.go [hunk 9]
```diff
@@ -76,14 +76,25 @@ type ChannelGraphSource interface {
...
+
+	// MarkEdgeZombie marks an edge as a zombie within our zombie index.
+	MarkEdgeZombie(chanID lnwire.ShortChannelID) error
+
+	// MarkEdgeLive clears an edge from our zombie index, deeming it as
+	// live.
+	MarkEdgeLive(chanID lnwire.ShortChannelID) error
+
+	// IsZombieEdge returns whether the edge is considered zombie.
+	IsZombieEdge(chanID lnwire.ShortChannelID) bool
```
[the remark this answers is not in the document]
Comment (wpaulino) [reply routing/router.go]: Slimmed it down a bit since the gossiper now only requires knowing of `MarkEdgeLive`.

File: discovery/gossiper.go (hunk 1 above)
Comment (wpaulino) [reply discovery/gossiper.go]: Fixed!

**Comment by wpaulino** [comment]
Comments addressed, PTAL @cfromknecht @Roasbeef.

File: routing/ann_validation.go [hunk 10]
```diff
@@ -132,20 +133,28 @@ func ValidateChannelUpdateAnn(pubKey *btcec.PublicKey, capacity btcutil.Amount,
...
+}
+
+// VerifyChannelUpdateSignature verifies that the channel update message was
+// signed by the party with the given node public key.
+func VerifyChannelUpdateSignature(msg *lnwire.ChannelUpdate,
+	pubKey *btcec.PublicKey) error {
+
+	data, err := msg.DataToSign()
 	if err != nil {
-		return errors.Errorf("unable to reconstruct message: %v", err)
```
Comment (cfromknecht) [inline routing/ann_validation.go]: 👍 one day, we will kill go-errors

File: channeldb/graph.go [hunk 11]
```diff
@@ -1631,11 +1664,21 @@ func delChannelByEdge(edges *bbolt.Bucket, edgeIndex *bbolt.Bucket,
...
-	return chanIndex.Delete(b.Bytes())
+	if err := chanIndex.Delete(b.Bytes()); err != nil {
+		return err
+	}
+
+	var pubKey1, pubKey2 [33]byte
+	copy(pubKey1[:], nodeKeys[:33])
+	copy(pubKey2[:], nodeKeys[33:])
+
+	return markEdgeZombie(
```
Comment (Roasbeef) [inline channeldb/graph.go]: We may want to distinguish between _deleting_ and edge, and marking it as a zombie...

Comment (wpaulino) [reply channeldb/graph.go]: Fixed.

File: channeldb/graph.go [hunk 12]
```diff
@@ -2782,6 +2793,107 @@ func (c *ChannelGraph) NewChannelEdgePolicy() *ChannelEdgePolicy {
...
+
+// markEdgeZombie marks an edge as a zombie within our zombie index. The public
+// keys should represent the node public keys of the two parties involved in the
+// edge.
+func markEdgeZombie(zombieIndex *bbolt.Bucket, chanID uint64, pubKey1,
+	pubKey2 [33]byte) error {
+
+	var k [8]byte
+	byteOrder.PutUint64(k[:], chanID)
+	v := make([]byte, 66)
```
Comment (halseth) [inline channeldb/graph.go]: `var v [66]byte`

File: channeldb/graph.go [hunk 13]
```diff
@@ -1638,12 +1653,29 @@ func delChannelByEdge(edges *bbolt.Bucket, edgeIndex *bbolt.Bucket,
...
 	}
-	return chanIndex.Delete(b.Bytes())
+	if err := chanIndex.Delete(b.Bytes()); err != nil {
+		return err
+	}
+
+	// Finally, we'll mark the edge as a zombie within our index if it's
+	// being removed due to the channel becoming a zombie. We do this to
+	// ensure we don't store unnecessary data for spent channels.
+	if !isZombie {
```
Comment (halseth) [inline channeldb/graph.go]: huh, missing param?

File: routing/router.go [hunk 14]
```diff
@@ -1009,11 +1009,14 @@ func (r *ChannelRouter) processUpdate(msg interface{}) error {
...
+		_, _, exists, isZombie, err := r.cfg.Graph.HasChannelEdge(
+			msg.ChannelID,
+		)
 		if err != nil && err != channeldb.ErrGraphNoEdgesFound {
 			return errors.Errorf("unable to check for edge "+
 				"existence: %v", err)
-		} else if exists {
+		}
+		if exists || isZombie {
 			return newErrf(ErrIgnored, "Ignoring msg for known "+
```
Comment (halseth) [inline routing/router.go]: Would be useful to return a slightly different msg for zombies.

File: routing/router.go [hunk 15]
```diff
@@ -1141,7 +1143,8 @@ func (r *ChannelRouter) processUpdate(msg interface{}) error {
 
 		// If the channel doesn't exist in our database, we cannot
 		// apply the updated policy.
-		if !exists {
+		isStaleUpdate := time.Since(msg.LastUpdate) > r.cfg.ChannelPruneExpiry
```
Comment (halseth) [inline routing/router.go]: This looks like new behavior (other than zombie check)?

I think it would be worth documenting what is going on here, returning different error messages (see comment above), and possibly make this its own commit (or describe in the commit message what this is good for).

File: discovery/gossiper.go [hunk 16]
```diff
@@ -1808,56 +1817,100 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
...
+				"graph(shortChanID=%v), saving for "+
+				"reprocessing later", shortChanID)
+
+			// NOTE: We don't return anything on the error channel
+			// for this message, as we expect that will be done when
+			// this ChannelUpdate is later reprocessed.
+			return nil
+
+		case nil:
+			// No error, fallthrough.
```
Comment (halseth) [inline discovery/gossiper.go]: you mean `break`? 😛

File: discovery/gossiper.go [hunk 17]
```diff
@@ -1808,56 +1817,100 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
...
-				d.rejectMtx.Lock()
-				d.recentRejects[msg.ShortChannelID.ToUint64()] = struct{}{}
-				d.rejectMtx.Unlock()
 				return nil
 			}
+
+			log.Debugf("Removed edge with chan_id=%v from zombie "+
+				"index", msg.ShortChannelID)
+
+			fallthrough
```
Comment (halseth) [inline discovery/gossiper.go]: Add a comment explaining why this falltrhoughs. Also update comment below?

File: discovery/gossiper.go [hunk 18]
```diff
@@ -1808,56 +1817,100 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
...
-				)
-				d.pChanUpdMtx.Unlock()
+		switch err {
+		case channeldb.ErrZombieEdge:
+			// If we've previously deemed the channel for this
+			// update a zombie and we came across a fresh update,
+			// then we'll mark it as live to ensure we can process
+			// any future updates for this edge until it is
+			// considered zombie again.
+			if time.Since(timestamp) > d.cfg.LiveEdgeHorizon {
```
Comment (halseth) [inline discovery/gossiper.go]: seems like this check is already done by `IsStaleEdgePolicy`

File: discovery/gossiper_test.go [hunk 19]
```diff
@@ -2191,6 +2193,255 @@ func TestForwardPrivateNodeAnnouncement(t *testing.T) {
...
+	// is a zombie or not.
+	processAnnouncement := func(ann lnwire.Message, isZombie, expectsErr bool) {
+		t.Helper()
+
+		errChan := ctx.gossiper.ProcessRemoteAnnouncement(
+			ann, remotePeer,
+		)
+		select {
+		case err := <-errChan:
+			if isZombie && (expectsErr && err == nil ||
```
Comment (halseth) [inline discovery/gossiper_test.go]: this is kinda hard to comprehend 😛  Better to divide into distinct cases I think

File: discovery/gossiper_test.go [hunk 20]
```diff
@@ -2191,6 +2193,255 @@ func TestForwardPrivateNodeAnnouncement(t *testing.T) {
...
+		if err != nil {
+			t.Fatalf("expected to process live channel update: %v",
+				err)
+		}
+	case <-time.After(time.Second):
+		t.Fatal("expected to process announcement")
+	}
+
+	select {
+	case <-ctx.broadcastedMessage:
```
Comment (halseth) [inline discovery/gossiper_test.go]: same, check expected message.

File: discovery/gossiper_test.go [hunk 21]
```diff
@@ -2191,6 +2193,255 @@ func TestForwardPrivateNodeAnnouncement(t *testing.T) {
...
+			}
+			if !isZombie && err != nil {
+				t.Fatalf("expected to process live channel "+
+					"message: %v", err)
+			}
+		case <-time.After(time.Second):
+			t.Fatal("expected to process announcement")
+		}
+		select {
+		case <-ctx.broadcastedMessage:
```
Comment (halseth) [inline discovery/gossiper_test.go]: should check that this is the expected message.

Comment (wpaulino) [reply discovery/gossiper_test.go]: Fixed.

File: discovery/gossiper_test.go (hunk 20 above)
Comment (wpaulino) [reply discovery/gossiper_test.go]: Fixed.

File: discovery/gossiper_test.go (hunk 19 above)
Comment (wpaulino) [reply discovery/gossiper_test.go]: Fixed.

File: discovery/gossiper.go (hunk 18 above)
Comment (wpaulino) [reply discovery/gossiper.go]: Fixed.

File: discovery/gossiper.go (hunk 17 above)
Comment (wpaulino) [reply discovery/gossiper.go]: Fixed.

File: discovery/gossiper.go (hunk 16 above)
Comment (wpaulino) [reply discovery/gossiper.go]: Fixed.

File: routing/router.go (hunk 15 above)
Comment (wpaulino) [reply routing/router.go]: Fixed.

File: routing/router.go (hunk 14 above)
Comment (wpaulino) [reply routing/router.go]: Fixed.

File: channeldb/graph.go (hunk 13 above)
Comment (wpaulino) [reply channeldb/graph.go]: Fixed.

File: channeldb/graph.go (hunk 12 above)
Comment (wpaulino) [reply channeldb/graph.go]: Fixed.

File: discovery/gossiper.go [hunk 22]
```diff
@@ -1808,56 +1809,99 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
...
+		case nil:
+			break
+
+		case channeldb.ErrZombieEdge:
+			// Since we've deemed the update as not stale above,
+			// before marking it live, we'll make sure it has been
+			// signed by the correct party. The least-significant
+			// bit in the flag on the channel update tells us which
+			// edge is being updated.
+			var pubKey *btcec.PublicKey
```
Comment (cfromknecht) [inline discovery/gossiper.go]: not a blocker, just noting that this same block of code exists in many places w/in the code base. would be great to have a method like
```golang
func (c *ChannelEdgeInfo) DirectionKey(flags lnwire.ChannelFlags) *btcec.PublicKey
```

File: discovery/gossiper.go [hunk 23]
```diff
@@ -1808,56 +1809,99 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(
...
+			d.prematureChannelUpdates[shortChanID] = append(
+				d.prematureChannelUpdates[shortChanID], nMsg,
+			)
+			d.pChanUpdMtx.Unlock()
+
+			log.Debugf("Got ChannelUpdate for edge not found in "+
+				"graph(shortChanID=%v), saving for "+
+				"reprocessing later", shortChanID)
+
+			// NOTE: We don't return anything on the error channel
```
Comment (halseth) [inline discovery/gossiper.go]: Can this potentially cause the gossiper to deadlock if the announcement never comes through? 🤔

Comment (wpaulino) [reply discovery/gossiper.go]: No because we return. The `errChan` in which we're sending the error back is given to the caller.

Comment (halseth) [reply discovery/gossiper.go]: Was just wondering who waits for the value on the channel, but looks like we never actually read it for remote annonucements...
